File: 66-plus-one/66-plus-one.js
/**
 * @param {number[]} digits
 * @return {number[]}
 */
var plusOne = function(digits) {
    let carry = 0
    let leftover = 0
    let i = digits.length - 1 //0
    let sum = digits[i] + 1 //10
    
    if (sum >= 10) {
        calculateCarryAndLeftover()
    } else {
        digits[i] = sum;
    }
    function calculateCarryAndLeftover() {
        leftover = sum - 10 //0
        carry = 1 //1
        carryHelp()
    }
    
    function carryHelp() {
        digits[i] = leftover //[0]
        i-- //-1
        if (i < 0) { 
            sum = carry //1
        } else {
            sum = digits[i] + carry   
        }
        if (sum >= 10) {
           calculateCarryAndLeftover()
        } else {
            if (i < 0) {
                digits.unshift(sum) //[1,0]
            } else {
                digits[i] = sum;            
            }
        }   
    }    
    return digits
};

File: 66-plus-one/66-plus-one.java
class Solution {
    public int[] plusOne(int[] digits) {
        int sum = 0;
        int carry = 0;
        
        for (int i = digits.length - 1; i >= 0; i--) {
           sum = digits[i] + 1;
           carry = 0;
           if (sum >= 10) {
               carry = 1;
               sum -= 10;
           }  
           digits[i] = sum;
           if (carry == 0) {
               break;
           }
       }
        if (carry == 1) {
            int[] newDigits = new int[digits.length + 1];
            newDigits[0] = 1;
            for (int i = 0; i < digits.length; i++) {
                newDigits[i+1] = digits[i];
                return newDigits;
            }
        }
       return digits;
    }
}

/*
class Solution {
    public int[] plusOne(int[] digits) {
        int carry = 0;
        int i = digits.length - 1;
        
        int sum = digits[i] + 1;
        if (sum >= 10) {
            carry = 1;
            while (carry == 1) {
                sum -=10;
                digits[i] = sum;
                i--;
                if (i < 0) break;
                sum = digits[i] + carry;
                carry = sum >=10 ? 1 : 0; 
            }
            
        } else { digits[i] = sum;}
         
        if (carry == 1) {
            int newDigits[] = new int[digits.length + 1];
            newDigits[0] = 1;
            for (int j = 0; j < digits.length; j++) {
                newDigits[j+1] = digits[j];
            }
            return newDigits;
        } 
        
        return digits;
    }
}
*/